src/librustc_mir/transform/const_prop.rs

```diff
@@ -890,9 +907,18 @@ impl<'mir, 'tcx> MutVisitor<'tcx> for ConstPropagator<'mir, 'tcx> {
                     );
                     Self::remove_const(&mut self.ecx, place.local);
                 }
+            } else {
+                trace!(
+                    "cannot propagate into {:?}, because the type of the local is generic.",
+                    place,
+                );
+                Self::remove_const(&mut self.ecx, place.local);
             }
         } else {
             match statement.kind {
+                StatementKind::SetDiscriminant { .. } => {
+                    self.use_ecx(|this| this.ecx.statement(statement));
```

Comment:
Found the issue... And the fact that this has taken me so long to find makes me think our `locals_of_current_block` scheme is too fragile. Maybe instead of collecting locals to nuke at the end of a block, we should just nuke all locals that are `OnlyInsideOwnBlock` at the end of each block. This is likely a bit more expensive if there are many locals, but we can measure that and figure things out from there.

Reply:
Opened https://github.com/rust-lang/rust/pull/73757 about that.